File: week09/7/[241101](Leet230)(Kth Smallest Element in a BST).py
"""문제
Given the root of a binary search tree, and an integer k, return the kth smallest value (1-indexed) of all the values of the nodes in the tree.
"""

"""문제풀이
    1. values 저장하는 list 생성
    2. dfs_inorder로 방문하며 만나는 노드 val 저장
    3. return val[k-1]
"""

# Definition for a binary tree node.
class TreeNode(object):
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right
        
class Solution():
    def dfs(self, curNode, values):
        if curNode is None:
            return values
        left_values = self.dfs(curNode.left, values) if curNode.left else []
        right_values = self.dfs(curNode.right, values) if curNode.right else []
        return left_values + [curNode.val] + right_values
            
    def kthSmallest(self, root, k):
        values = self.dfs(root, [])
        return values[k-1]
    

## ------ 아래는 따로 values list를 저장하지 않고 진행하는 방법.

#보다 효율적인 코드 : count가 k에 도달하면 중단.
    ## inorder traversal 중 k번째 작은 요소 찾으면 그 값이 self.result에 저장됨. 이후 재귀적으로 다른 노드 방문하기 전, 이 조건을 검사해서 더이상의 탐색을 하지 않도록 함. 
class Solution:
    def kthSmallest(self, root, k):
        self.count = 0
        self.result = None
        
        def inorder(node):
            if not node or self.result is not None:
                return
            
            inorder(node.left)
            self.count += 1
            if self.count == k:
                self.result = node.val
                return
            inorder(node.right)
        
        inorder(root)
        return self.result
    

# class 변수 없이 구현 (중위 순회 동안 count가 k에 도달하면 해당 노드의 값이 result[0]에 저장되고, 이후의 모든 순회는 조기에 종료)
class Solution:
    def kthSmallest(self, root, k):
        def inorder(node, count, result):
            if not node or result[0] is not None:
                return count, result
            
            # 왼쪽 자식 노드 방문
            count, result = inorder(node.left, count, result)
            if result[0] is not None:
                return count, result
            
            # 현재 노드 처리
            count += 1
            if count == k:
                result[0] = node.val
                return count, result
            
            # 오른쪽 자식 노드 방문
            return inorder(node.right, count, result)
        
        # count 초기화, result를 리스트로 초기화 (참조 유지를 위해)
        result = [None]
        count = 0
        _, result = inorder(root, count, result)
        return result[0]

File: week09/7/[241101](Leet230)(Kth Smallest Element in a BST).c
/*문제
Given the root of a binary search tree, and an integer k, return the kth smallest value (1-indexed) of all the values of the nodes in the tree.*/

/*Constraints:
    The number of nodes in the tree is n.
    1 <= k <= n <= 104
    0 <= Node.val <= 104
*/

// 방법 1 : values list 활용

/* 방법 2 : dfs동안 k값을 줄이면서 k==0 시점의 node.val 반환
    - 유의점! : 자기자신을 방문하면 curCount--.  그리고 난 후! curCount ==0 인지 확인.
*/


// 방법 1 ---------------------------------------
#include <stdio.h>
#include <stdlib.h>

//Definition for a binary tree node.
struct TreeNode {
    int val;
    struct TreeNode *left;
    struct TreeNode *right;
};

void sizeUp(int** values, int* sizeValues, int* valuesCapacity) {  //포인터인 values(int*형)의 자체의 값을 변경해야 함. 그러려면 int* values의 주소에 접근해야 하므로..  int**로 받아와야 함에 주의. 
    if (*valuesCapacity == *sizeValues) {
        *valuesCapacity *= 2;
        int* temp = realloc(*values, sizeof(int) * (*valuesCapacity));
        if (temp == NULL) {
            // 메모리 할당 실패 처리
            fprintf(stderr, "Memory allocation failed.\n");
            exit(EXIT_FAILURE);
        }
        *values = temp;  // 주소를 업데이트
    } else {
        return;
    }
}
void dfs(struct TreeNode* curNode, int** values, int* sizeValues, int* valuesCapacity, int k){
    if (curNode == NULL) return;
    // if (*sizeValues > k) return;

    if (curNode->left) dfs(curNode->left, values, sizeValues, valuesCapacity,k);
    sizeUp(values, sizeValues, valuesCapacity); // dfs에서 int* values로 argument 주고, 여기서 &values라고 전달한다면, realloc이 제대로 되지 않음. 왜냐면 함수 호출할 때 포인터의 복사본이 만들어지기 때문에 그 주솟값이 다름.
    (*sizeValues)++;
    (*values)[(*sizeValues)-1] = curNode->val;
    if (curNode->right) dfs(curNode->right, values, sizeValues,valuesCapacity, k); 
}


int kthSmallest(struct TreeNode* root, int k){
    int valuesCapacity = k; // 초기에 그냥 적당한 크기로 줌.
    int* values = (int*)malloc(sizeof(int) * (valuesCapacity));
    int sizeValues = 0;

    dfs(root, &values, &sizeValues, &valuesCapacity, k);

    int answer = values[k-1];

    free(values);
    return answer;
}



//방법2 : 통과 ----------------------------
#include <stdio.h>
#include <stdlib.h>

//Definition for a binary tree node.
struct TreeNode {
    int val;
    struct TreeNode *left;
    struct TreeNode *right;
};

void dfs(struct TreeNode* curNode, int* curCount, int* answer){
    if (curNode == NULL) return;

    if (curNode->left) dfs(curNode->left, curCount, answer);
    (*curCount)--;   #빼는 게 먼저 선행되어야 한다. *curCount ==0 여부를 확인하기 전에.
    if ((*curCount) == 0){
        *answer = curNode->val;
        return;
    }
    if (curNode->right) dfs(curNode->right, curCount, answer);

}

int kthSmallest(struct TreeNode* root, int k){
    int curCount = k;
    int answer = 0;

    dfs(root, &curCount, &answer);

    return answer;
}
